waspc/data/Generator/templates/react-app/src/router.tsx

```diff
@@ -25,35 +25,31 @@ export const routeNameToRouteComponent = {
   {=/ routes =}
 } as const;
 
-const router = (
-  <Router basename="{= baseDir =}">
-    {=# rootComponent.isDefined =}
-    <{= rootComponent.importIdentifier =}>
-    {=/ rootComponent.isDefined =}
-    <Switch>
-      {=# isExternalAuthEnabled =}
-      {/* 
-        Wasp specific routes *must* go first to prevent user
-        defined routes from overriding them.
-        Details in https://github.com/wasp-lang/wasp/issues/2029
-      */}
-      <Route exact path="{= oAuthCallbackPath =}">
-        <OAuthCallbackPage />
-      </Route>
-      {=/ isExternalAuthEnabled =}
-      {Object.entries(routes).map(([routeKey, route]) => (
-        <Route
-          exact
-          key={routeKey}
-          path={route.to}
-          component={routeNameToRouteComponent[routeKey]}
-        />
-      ))}
-    </Switch>
-    {=# rootComponent.isDefined =}
-    </{= rootComponent.importIdentifier =}>
-    {=/ rootComponent.isDefined =}
-  </Router>
-)
-
-export default router
+const waspDefinedRoutes = [
+  {=# isExternalAuthEnabled =}
+  {
+    path: "{= oAuthCallbackPath =}",
+    Component: OAuthCallbackPage,
+  },
+  {=/ isExternalAuthEnabled =}
+]
+const userDefinedRoutes = Object.entries(routes).map(([routeKey, route]) => {
+  return {
+    path: route.to,
+    Component: routeNameToRouteComponent[routeKey],
+  }
+})
+
+const browserRouter = createBrowserRouter([{
+  path: '/',
+  {=# rootComponent.isDefined =}
+  element: <{= rootComponent.importIdentifier =} />,
+  {=/ rootComponent.isDefined =}
+  children: [
+    ...waspDefinedRoutes,
+    ...userDefinedRoutes,
+  ],
+}])
+
+
+export const router = <RouterProvider router={browserRouter} />
```

Comment:
From the docs:
> You should only opt into the Component API for data routes via RouterProvider. Using this API on a <Route> inside <Routes> will de-optimize React's ability to reuse the created element across renders.

Since we are using `Component` and not element API, using the `RouterProvider` is optimal.